src/fsharp/CheckFormatStrings.fs

```diff
@@ -47,31 +49,153 @@ let newInfo () =
     addZeros       = false
     precision      = false}
 
-let parseFormatStringInternal (m:range) (g: TcGlobals) (context: FormatStringCheckContext option) fmt bty cty = 
-    // Offset is used to adjust ranges depending on whether input string is regular, verbatim or triple-quote.
-    // We construct a new 'fmt' string since the current 'fmt' string doesn't distinguish between "\n" and escaped "\\n".
-    let (offset, fmt) = 
+let parseFormatStringInternal (m: range) (fragRanges: range list) (g: TcGlobals) isInterpolated isFormattableString (context: FormatStringCheckContext option) fmt printerArgTy printerResidueTy = 
+
+    // As background: the F# compiler tokenizes strings on the assumption that the only thing you need from
+    // them is the actual corresponding text, e.g. of a string literal.  This means many different textual input strings
+    // in the input file correspond to the 'fmt' string we have here.
+    //
+    // The problem with this is that when we go to colorize the format specifiers in string, we need to do
+    // that with respect to the original string source text in order to lay down accurate colorizations.
+    //
+    // One approach would be to change the F# lexer to also crack every string in a more structured way, recording
+    // both the original source text and the actual string literal.  However this would be invasive and possibly
+    // too expensive since the vast majority of strings don't need this treatment.
+    //
+    // So instead, for format strings alone - and only when processing in the IDE - we crack the "original"
+    // source of the string by going back and getting the format string from the input source file by using the
+    // relevant ranges
+    //
+    // For interpolated strings this may involve many fragments, e.g. 
+    //     $"abc %d{"
+    //     "} def %s{"
+    //     "} xyz"
+    // In this case we are given the range of each fragment. One annoying thing is that we must lop off the
+    // quotations, $, {, } symbols off the end of each string fragment. This information should probably
+    // be given to us by the lexer.
+    //
+    // Note this also means that when compiling (command-line or background IncrementalBuilder in the IDE
+    // there are no accurate intra-string ranges available for exact error message locations within the string.
+    // The 'm' range passed as an input is however accurate and covers the whole string.
+    ///
+    let fmt, fragments = 
+
+        //printfn "--------------------" 
+        //printfn "context.IsSome = %b" context.IsSome
+        //printfn "fmt  = <<<%s>>>" fmt
+        //printfn "isInterpolated = %b" isInterpolated
+        //printfn "fragRanges = %A" fragRanges
+
         match context with
-        | Some context ->
+        | Some context when fragRanges.Length > 0 ->
             let sourceText = context.SourceText
+            //printfn "sourceText.IsSome = %b" sourceText.IsSome
```

Comment:
Can these be removed?

Reply:
I think I'll leave these in for now as they were so crucial to debugging issues in here quickly and accurately. We sometimes do that, it's ok